File: dotnet/src/Microsoft.Agents.Validation/Validators/MessageValidators/ChannelMessageValidator.cs
using Microsoft.Agents.Abstractions.Models;

namespace Microsoft.Agents.Validation.Validators.MessageValidators;

/// <summary>
/// Validator for ChannelMessage.
/// </summary>
public class ChannelMessageValidator : MessageValidatorBase<ChannelMessage>
{
    protected override List<ValidationError> ValidateSpecificFields(ChannelMessage message, ValidationContext? context)
    {
        var errors = new List<ValidationError>();

        // ROLE-006: ChannelMessage can contain EventContent, TraceContent, ActionContent
        if (message.Contents != null && message.Contents.Count > 0)
        {
            var invalidContent = message.Contents
                .Where(c => c is not EventContent && c is not TraceContent && c is not ActionContent)
                .ToList();

            if (invalidContent.Any())
            {
                var contentTypes = string.Join(", ", invalidContent.Select(c => c.GetType().Name));
                errors.Add(CreateError(
                    ValidationErrorCode.ROLE_006,
                    $"ChannelMessage can only contain EventContent, TraceContent, or ActionContent, found: {contentTypes}",
                    "Contents"));
            }

            // Validate content items
            errors.AddRange(ValidateContentItems(message.Contents, context));
        }

        return errors;
    }
}

File: typescript/packages/agents-protocol-validation/src/validators/message/ChannelMessageValidator.ts
import type { ChannelMessage } from '@microsoft/agents-protocol-abstractions';
import type { ValidationContext } from '../../core/ValidationContext';
import { MessageValidatorBase } from '../base/MessageValidatorBase';
import { ValidationError } from '../../core/ValidationError';
import { ValidationErrorCode } from '../../core/ValidationErrorCode';

/**
 * Validator for ChannelMessage.
 */
export class ChannelMessageValidator extends MessageValidatorBase<ChannelMessage> {
  protected validateSpecificFields(message: ChannelMessage, context?: ValidationContext): ValidationError[] {
    const errors: ValidationError[] = [];

    // Channel-specific validation can be added here

    return errors;
  }

  protected getAllowedContentTypes(): string[] {
    // ROLE-006: Channel messages can contain UI and system content types
    return [
      'text',
      'typing-indicator',
      'message-reaction',
      'message-delete',
      'message-update',
      'suggested-actions',
      'action',
      'user-input-request',
      'event',
      'trace',
      'content-filter-result',
    ];
  }

  protected getRoleSpecificErrorCode(): string {
    return ValidationErrorCode.ROLE_006;
  }

  protected getMessageTypeName(): string {
    return 'ChannelMessage';
  }

  protected requiresContent(): boolean {
    return true;
  }
}
